README.md
# wordify

A Clojure library designed to convert numbers into the language word equivalents 
(i.e 1024 = "one thousand and twenty four").

## Usage

Example:

```
(wordify/wordify-number 50472)
=> "fifty thousand four hundred and seventy two"

(wordify/wordify-number "50472")
=> "fifty thousand four hundred and seventy two"

(wordify/wordify-number 50.472)
=> "fifty and four hundred and seventy two thousandths"
```

## Installation

Build the jar file and add it to the local repo:
```
lein install
```

Add the following to your project :dependencies:
```
[wordify "0.1.0"]
```

And add to your namespace:
```
(:require [wordify.core :as wordify])
```

## License

Copyright © 2021 Russell Davey

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-04-24
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-04-24
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/wordify/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/wordify/compare/0.1.0...0.1.1


test/wordify/data/test_data.clj
(ns wordify.data.test-data)

(def numbers-as-words
  ["zero"
   "one"
   "two"
   "three"
   "four"
   "five"
   "six"
   "seven"
   "eight"
   "nine"
   "ten"
   "eleven"
   "twelve"
   "thirteen"
   "fourteen"
   "fifteen"
   "sixteen"
   "seventeen"
   "eighteen"
   "nineteen"
   "twenty"
   "twenty one"
   "twenty two"
   "twenty three"
   "twenty four"
   "twenty five"
   "twenty six"
   "twenty seven"
   "twenty eight"
   "twenty nine"
   "thirty"
   "thirty one"
   "thirty two"
   "thirty three"
   "thirty four"
   "thirty five"
   "thirty six"
   "thirty seven"
   "thirty eight"
   "thirty nine"
   "forty"
   "forty one"
   "forty two"
   "forty three"
   "forty four"
   "forty five"
   "forty six"
   "forty seven"
   "forty eight"
   "forty nine"
   "fifty"
   "fifty one"
   "fifty two"
   "fifty three"
   "fifty four"
   "fifty five"
   "fifty six"
   "fifty seven"
   "fifty eight"
   "fifty nine"
   "sixty"
   "sixty one"
   "sixty two"
   "sixty three"
   "sixty four"
   "sixty five"
   "sixty six"
   "sixty seven"
   "sixty eight"
   "sixty nine"
   "seventy"
   "seventy one"
   "seventy two"
   "seventy three"
   "seventy four"
   "seventy five"
   "seventy six"
   "seventy seven"
   "seventy eight"
   "seventy nine"
   "eighty"
   "eighty one"
   "eighty two"
   "eighty three"
   "eighty four"
   "eighty five"
   "eighty six"
   "eighty seven"
   "eighty eight"
   "eighty nine"
   "ninety"
   "ninety one"
   "ninety two"
   "ninety three"
   "ninety four"
   "ninety five"
   "ninety six"
   "ninety seven"
   "ninety eight"
   "ninety nine"])


test/wordify/number_wordify_test.clj
(ns wordify.number-wordify-test
  (:require [clojure.test :refer :all]
            [wordify.number-wordify :as target]
            [wordify.data.test-data :as td]))

(deftest negative-numbers-test
  (let [results (target/int-number->words -101)]
    (is (= "negative one hundred and one" results))))

(deftest lower-number-test
  (let [test-numbers (range 0 21)
        results (map target/int-number->words test-numbers)]
    (is (= (take 21 td/numbers-as-words) results))))

(deftest higher-number-test
  (let [test-numbers (range 21 100)
        results (map target/int-number->words test-numbers)]
    (is (= (drop 21 td/numbers-as-words) results))))

(deftest large-number-test
  (testing "hundreds"
    (let [test-numbers [105 123 500]
          results (map target/int-number->words test-numbers)]
      (is (= "one hundred and five" (first results)))
      (is (= "one hundred and twenty three" (second results)))
      (is (= "five hundred" (nth results 2)))))

  (testing "thousands"
    (let [test-numbers [1005 1042 1105]
          results (map target/int-number->words test-numbers)]
      (is (= "one thousand and five" (first results)))
      (is (= "one thousand and forty two" (second results)))
      (is (= "one thousand one hundred and five" (nth results 2)))))

  (testing "tens of thousands"
    (let [test-numbers [10005 50123]
          results (map target/int-number->words test-numbers)]
      (is (= "ten thousand and five" (first results)))
      (is (= "fifty thousand one hundred and twenty three" (second results)))))

  (testing "hundreds of thousands"
    (let [test-numbers [112345 500000 500123]
          results (map target/int-number->words test-numbers)]
      (is (= "one hundred and twelve thousand three hundred and forty five" (first results)))
      (is (= "five hundred thousand" (second results)))
      (is (= "five hundred thousand one hundred and twenty three" (nth results 2)))))

  (testing "millions"
    (let [test-numbers [56945781 999999999]
          results (map target/int-number->words test-numbers)]
      (is (= "fifty six million nine hundred and forty five thousand seven hundred and eighty one" (first results)))
      (is (= "nine hundred and ninety nine million nine hundred and ninety nine thousand nine hundred and ninety nine" (second results))))))

(deftest extra-large-number-test
  (testing "billions"
    (let [test-numbers [7439647773 74396477730]
          results (map target/int-number->words test-numbers)]
      (is (= "seven billion four hundred and thirty nine million six hundred and forty seven thousand seven hundred and seventy three" (first results)))
      (is (= "seventy four billion three hundred and ninety six million four hundred and seventy seven thousand seven hundred and thirty" (second results)))))

  (testing "quadrillion"
    (let [test-numbers [9999999999999999]
          results (map target/int-number->words test-numbers)]
      (is (= "nine quadrillion nine hundred and ninety nine trillion nine hundred and ninety nine billion nine hundred and ninety nine million nine hundred and ninety nine thousand nine hundred and ninety nine" (first results)))))

  (testing "quintillion"
    (let [test-numbers [9999999999999999999]
          results (map target/int-number->words test-numbers)]
      (is (= "nine quintillion nine hundred and ninety nine quadrillion nine hundred and ninety nine trillion nine hundred and ninety nine billion nine hundred and ninety nine million nine hundred and ninety nine thousand nine hundred and ninety nine" (first results)))))

  (testing "tredecillion"
    (let [test-numbers (bigint (apply str (repeat 45 "9")))
          results (target/int-number->words test-numbers)]
      (is (= "nine hundred and ninety nine tredecillion nine hundred and ninety nine duodecillion nine hundred and ninety nine undecillion nine hundred and ninety nine decillion nine hundred and ninety nine nonillion nine hundred and ninety nine octillion nine hundred and ninety nine septillion nine hundred and ninety nine sextillion nine hundred and ninety nine quintillion nine hundred and ninety nine quadrillion nine hundred and ninety nine trillion nine hundred and ninety nine billion nine hundred and ninety nine million nine hundred and ninety nine thousand nine hundred and ninety nine"
             results))))

  (testing "vigintillion"
    (let [test-numbers (bigint (str "1" (apply str (repeat 65 "0"))))
          results (target/int-number->words test-numbers)]
      (is (= "one hundred vigintillion"
             results)))))

(deftest numbers-as-strings)

(deftest numbers-as-doubles
  (let [test-numbers [1.2 1.23 5.6 567.232]
        results (map target/double-number->words test-numbers)]
    (is (= "one and two tenths" (first results)))
    (is (= "one and twenty three hundredths" (second results)))
    (is (= "five and six tenths" (nth results 2)))
    (is (= "five hundred and sixty seven and two hundred and thirty two thousandths" (nth results 3)))))

(deftest unhappy-paths
  ; given a string returns nil
  (is (= nil (target/int-number->words "1")))
  ; given a string without numbers returns nil
  (is (= nil (target/int-number->words "dsfdsfds")))
  ; given a keyword returns nil
  (is (= nil (target/int-number->words :34)))
  ; given a float returns nil
  (is (= nil (target/int-number->words 100.03434)))
  ; given the max number, does not return nil
  (is (not (nil? (target/int-number->words 7439647657730540004))))
  ; given a number above max returns nil
  (is (= nil (target/int-number->words (bigint (apply str (repeat 67 "9")))))))

test/wordify/conversion_utils_test.clj
(ns wordify.conversion-utils-test
  (:require [clojure.test :refer :all]
            [wordify.conversion-utils :as target]))

(deftest parse-string-into-integers
  (testing "basic"
    (let [result (target/safe-parse-int "343243243")]
      (is (= 343243243 result)))))

(deftest parse-strings-into-big-int
  (let [result (target/safe-parse-big-int "9999999999999999999")]
    (is (= 9999999999999999999 result))))

src/wordify_numbers/number_stringify.clj
(ns wordify-numbers.number-stringify
  (:require [clojure.string :as string]))

(def ^:private lower-numbers {"0" "zero"
                              "1" "one"
                              "2" "two"
                              "3" "three"
                              "4" "four"
                              "5" "five"
                              "6" "six"
                              "7" "seven"
                              "8" "eight"
                              "9" "nine"
                              "10" "ten"
                              "11" "eleven"
                              "12" "twelve"
                              "13" "thirteen"
                              "14" "fourteen"
                              "15" "fifteen"
                              "16" "sixteen"
                              "17" "seventeen"
                              "18" "eighteen"
                              "19" "nineteen"
                              "20" "twenty"})

(def ^:private higher-numbers {"2" "twenty"
                               "3" "thirty"
                               "4" "forty"
                               "5" "fifty"
                               "6" "sixty"
                               "7" "seventy"
                               "8" "eighty"
                               "9" "ninety"})

(def ^:private large-numbers {18 "quintillion"
                              15 "quadrillion"
                              12 "trillion"
                              9 "billion"
                              6 "million"
                              3 "thousand"
                              2 "hundred"})

(defn- safe-parse-long
  "given a string attempts to parse it as an integer"
  [str]
  (try
    (Long/parseLong str)
    (catch NumberFormatException _
      nil)))

(defn- between-zero-and-one-hundred?
  [i]
  (when i
    (and (> i 0) (< i 100))))

(defn- get-order-of-magnitude
  "given an integer convert to a base10 integer and return the matching magnitude and large number string
  from the large-numbers map"
  [i]
  (when-not (zero? i)
    (let [magnitude (int (Math/log10 i))]
      (first (filter (fn [[large-number-int _]]
                       (>= magnitude large-number-int))
                     large-numbers)))))

(defn- has-no-remainder?
  "given an integer divides by 10, returns true if there are no remainders and false if there are"
  [i] (zero? (rem i 10)))

(defn- within-acceptable-params?
  "given a number, validates that it is within acceptable parameters (i.e. is an int)"
  [number]
  (and (int? number)
       (> number -999999999999998)
       (< number 999999999999998)))

(defn int->words
  [i]
  (when (within-acceptable-params? i)
    (-> (let [s (str i)
              str-len (count s)]
          (cond
            (< i 0) (str "negative "
                         (int->words (- i)))
            (<= i 20) (lower-numbers s)
            (< i 100) (let [first-word (subs s 0 1)
                            second-word (subs s 1 2)]
                        (if (has-no-remainder? i)
                          (higher-numbers first-word)
                          (str (higher-numbers first-word) " " (lower-numbers second-word))))
            :else (let [[magnitude large-number-str] (get-order-of-magnitude i)
                        offset (- str-len magnitude)
                        first-word (-> (subs s 0 offset)
                                       safe-parse-long)
                        rest-words-int (-> (subs s offset str-len)
                                           safe-parse-long)
                        rest-words (when-not (zero? rest-words-int)
                                     rest-words-int)
                        conjunctive (if (between-zero-and-one-hundred? rest-words)
                                      " and "
                                      " ")]
                    (if (has-no-remainder? i)
                      (str (int->words first-word) " " large-number-str)
                      (str (int->words first-word)
                           " "
                           large-number-str
                           conjunctive
                           (int->words rest-words))))))
        string/trim)))

src/wordify/number_wordify.clj
(ns wordify.number-wordify
  (:require [clojure.string :as string]
            [wordify.conversion-utils :refer [safe-parse-int safe-parse-big-int]])
  (:import (clojure.lang BigInt)))

(def ^:private lower-numbers {"0" "zero"
                              "1" "one"
                              "2" "two"
                              "3" "three"
                              "4" "four"
                              "5" "five"
                              "6" "six"
                              "7" "seven"
                              "8" "eight"
                              "9" "nine"
                              "10" "ten"
                              "11" "eleven"
                              "12" "twelve"
                              "13" "thirteen"
                              "14" "fourteen"
                              "15" "fifteen"
                              "16" "sixteen"
                              "17" "seventeen"
                              "18" "eighteen"
                              "19" "nineteen"
                              "20" "twenty"})

(def ^:private higher-numbers {"2" "twenty"
                               "3" "thirty"
                               "4" "forty"
                               "5" "fifty"
                               "6" "sixty"
                               "7" "seventy"
                               "8" "eighty"
                               "9" "ninety"})

(def ^:private large-numbers {63 "vigintillion"
                              60 "novemdecillion"
                              57 "octodecillion"
                              54 "septendecillion"
                              51 "sexdecillion"
                              48 "quindecillion"
                              45 "quattuordecillion"
                              42 "tredecillion"
                              39 "duodecillion"
                              36 "undecillion"
                              33 "decillion"
                              30 "nonillion"
                              27 "octillion"
                              24 "septillion"
                              21 "sextillion"
                              18 "quintillion"
                              15 "quadrillion"
                              12 "trillion"
                              9 "billion"
                              6 "million"
                              3 "thousand"
                              2 "hundred"})

(def ^:private large-numbers-extended
  (reduce (fn [new-map [mag-int mag]]
            (if (> mag-int 2)
              (merge new-map {(+ 1 mag-int) (str "ten-" mag)
                              (+ 2 mag-int) (str "hundred-" mag)})
              new-map))
          (merge large-numbers {1 "ten"})
          large-numbers))

(defn- between-zero-and-one-hundred?
  [i]
  (when i
    (and (> i 0) (< i 100))))

(defn- get-magnitude-number
  [n]
  (-> n str count (- 1)))

(defn- get-order-of-magnitude
  "given an integer convert to a base10 integer and return the matching magnitude and large number string
  from the large-numbers map"
  ([i]
   (get-order-of-magnitude i large-numbers))
  ([i number-map]
   (when-not (zero? i)
     (let [magnitude (get-magnitude-number i)]
       (first (-> (filter (fn [[large-number-int _]]
                            (>= magnitude large-number-int))
                          number-map)
                  sort
                  reverse))))))

(defn- has-no-remainder?
  "given an integer divides by 10, returns true if there are no remainders and false if there are"
  [i] (zero? (rem i 10)))

(defn- within-acceptable-params?
  "given a number, validates that it is within acceptable parameters (i.e. is an int)"
  [number]
  (let [int-or-big-int (or (int? number)
                           (instance? BigInt number))
        within-range (<= (get-magnitude-number number) (+ (apply max (keys large-numbers)) 2))]
    (every? true? [int-or-big-int within-range])))

(defn- int->words
  [i]
  (let [s (str i)
        str-len (count s)]
    (cond
      (< i 0) (str "negative "
                   (int->words (- i)))
      (<= i 20) (lower-numbers s)
      (< i 100) (let [first-word (subs s 0 1)
                      second-word (subs s 1 2)]
                  (if (has-no-remainder? i)
                    (higher-numbers first-word)
                    (str (higher-numbers first-word) " " (lower-numbers second-word))))
      :else (let [[magnitude large-number-str] (get-order-of-magnitude i)
                  offset (- str-len magnitude)
                  first-word (-> (subs s 0 offset)
                                 safe-parse-int)
                  rest-words (-> (subs s offset str-len)
                                 safe-parse-big-int)
                  conjunctive (if (between-zero-and-one-hundred? rest-words)
                                " and "
                                " ")]
              (when-not (= i rest-words)
                (if (zero? rest-words)
                  (str (int->words first-word) " " large-number-str)
                  (str (int->words first-word)
                       " "
                       large-number-str
                       conjunctive
                       (int->words rest-words))))))))

(defn int-number->words
  [i]
  (when (within-acceptable-params? i)
    (int->words i)))

(defn string-number->words
  [i]
  (when-let [number (safe-parse-big-int i)]
    (int-number->words number)))

(defn double-number->words
  [i]
  (when-let [[int-number dec-number] (clojure.string/split (str i) #"\.")]
    (let [[_ magnitude] (get-order-of-magnitude (-> dec-number safe-parse-big-int (* 10))
                                                large-numbers-extended)]
      (str (string-number->words int-number)
           " and "
           (string-number->words dec-number)
           (when magnitude
             (str " " magnitude "ths"))))))